Build: `make -j$(nproc)` fails on first build

### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
$ uname -srvmo; cat /etc/os-release || lsb_release -a
Linux 6.1.7-200.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 18 17:11:49 UTC 2023 x86_64 GNU/Linux
NAME="Fedora Linux"
VERSION="37 (Workstation Edition)"
ID=fedora
VERSION_ID=37
VERSION_CODENAME=""
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Linux 37 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f37/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=37
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=37
SUPPORT_END=2023-11-14
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```


### Version

```sh
./k0s version     
v1.26.0+k0s.0-91-gc3d7c1b32
```

### Sysinfo

```sh
$ /k0s sysinfo
Machine ID: "6085114bbae18d873810a7c9d2e085be46818b4fc30282364ced5a6188a93dee" (from machine) (pass)
Total memory: 62.7 GiB (pass)
Disk space available for /var/lib/k0s: 851.1 GiB (pass)
Operating system: Linux (pass)
  Linux kernel release: 6.1.7-200.fc37.x86_64 (pass)
  Max. file descriptors per process: current: 524288 / max: 524288 (pass)
  Executable in path: modprobe: /usr/sbin/modprobe (pass)
  /proc file system: mounted (0x9fa0) (pass)
  Control Groups: version 2 (pass)
    cgroup controller "cpu": available (pass)
    cgroup controller "cpuacct": available (via cpu in version 2) (pass)
    cgroup controller "cpuset": available (pass)
    cgroup controller "memory": available (pass)
    cgroup controller "devices": available (assumed) (pass)
    cgroup controller "freezer": available (assumed) (pass)
    cgroup controller "pids": available (pass)
    cgroup controller "hugetlb": available (pass)
    cgroup controller "blkio": available (via io in version 2) (pass)
  CONFIG_CGROUPS: Control Group support: built-in (pass)
    CONFIG_CGROUP_FREEZER: Freezer cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_PIDS: PIDs cgroup subsystem: built-in (pass)
    CONFIG_CGROUP_DEVICE: Device controller for cgroups: built-in (pass)
    CONFIG_CPUSETS: Cpuset support: built-in (pass)
    CONFIG_CGROUP_CPUACCT: Simple CPU accounting cgroup subsystem: built-in (pass)
    CONFIG_MEMCG: Memory Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_HUGETLB: HugeTLB Resource Controller for Control Groups: built-in (pass)
    CONFIG_CGROUP_SCHED: Group CPU scheduler: built-in (pass)
      CONFIG_FAIR_GROUP_SCHED: Group scheduling for SCHED_OTHER: built-in (pass)
        CONFIG_CFS_BANDWIDTH: CPU bandwidth provisioning for FAIR_GROUP_SCHED: built-in (pass)
    CONFIG_BLK_CGROUP: Block IO controller: built-in (pass)
  CONFIG_NAMESPACES: Namespaces support: built-in (pass)
    CONFIG_UTS_NS: UTS namespace: built-in (pass)
    CONFIG_IPC_NS: IPC namespace: built-in (pass)
    CONFIG_PID_NS: PID namespace: built-in (pass)
    CONFIG_NET_NS: Network namespace: built-in (pass)
  CONFIG_NET: Networking support: built-in (pass)
    CONFIG_INET: TCP/IP networking: built-in (pass)
      CONFIG_IPV6: The IPv6 protocol: built-in (pass)
    CONFIG_NETFILTER: Network packet filtering framework (Netfilter): built-in (pass)
      CONFIG_NETFILTER_ADVANCED: Advanced netfilter configuration: built-in (pass)
      CONFIG_NF_CONNTRACK: Netfilter connection tracking support: module (pass)
      CONFIG_NETFILTER_XTABLES: Netfilter Xtables support: built-in (pass)
        CONFIG_NETFILTER_XT_TARGET_REDIRECT: REDIRECT target support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_COMMENT: "comment" match support: module (pass)
        CONFIG_NETFILTER_XT_MARK: nfmark target and match support: module (pass)
        CONFIG_NETFILTER_XT_SET: set target and match support: module (pass)
        CONFIG_NETFILTER_XT_TARGET_MASQUERADE: MASQUERADE target support: module (pass)
        CONFIG_NETFILTER_XT_NAT: "SNAT and DNAT" targets support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: "addrtype" address type match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_CONNTRACK: "conntrack" connection tracking match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_MULTIPORT: "multiport" Multiple port match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_RECENT: "recent" match support: module (pass)
        CONFIG_NETFILTER_XT_MATCH_STATISTIC: "statistic" match support: module (pass)
      CONFIG_NETFILTER_NETLINK: module (pass)
      CONFIG_NF_NAT: module (pass)
      CONFIG_IP_SET: IP set support: module (pass)
        CONFIG_IP_SET_HASH_IP: hash:ip set support: module (pass)
        CONFIG_IP_SET_HASH_NET: hash:net set support: module (pass)
      CONFIG_IP_VS: IP virtual server support: module (pass)
        CONFIG_IP_VS_NFCT: Netfilter connection tracking: built-in (pass)
      CONFIG_NF_CONNTRACK_IPV4: IPv4 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_REJECT_IPV4: IPv4 packet rejection: module (pass)
      CONFIG_NF_NAT_IPV4: IPv4 NAT: unknown (warning)
      CONFIG_IP_NF_IPTABLES: IP tables support: module (pass)
        CONFIG_IP_NF_FILTER: Packet filtering: module (pass)
          CONFIG_IP_NF_TARGET_REJECT: REJECT target support: module (pass)
        CONFIG_IP_NF_NAT: iptables NAT support: module (pass)
        CONFIG_IP_NF_MANGLE: Packet mangling: module (pass)
      CONFIG_NF_DEFRAG_IPV4: module (pass)
      CONFIG_NF_CONNTRACK_IPV6: IPv6 connetion tracking support (required for NAT): unknown (warning)
      CONFIG_NF_NAT_IPV6: IPv6 NAT: unknown (warning)
      CONFIG_IP6_NF_IPTABLES: IP6 tables support: module (pass)
        CONFIG_IP6_NF_FILTER: Packet filtering: module (pass)
        CONFIG_IP6_NF_MANGLE: Packet mangling: module (pass)
        CONFIG_IP6_NF_NAT: ip6tables NAT support: module (pass)
      CONFIG_NF_DEFRAG_IPV6: module (pass)
    CONFIG_BRIDGE: 802.1d Ethernet Bridging: module (pass)
      CONFIG_LLC: module (pass)
      CONFIG_STP: module (pass)
  CONFIG_EXT4_FS: The Extended 4 (ext4) filesystem: built-in (pass)
  CONFIG_PROC_FS: /proc file system support: built-in (pass
```

### What happened?

`make -j$(nproc)` fails on first, and possibly second build.

```sh
$ git rev-parse HEAD        
c3d7c1b32f1af461192ad5ad9689c8f74d8fe71f
```

### Steps to reproduce

1. Checkout `main` at `c3d7c1b32f1af461192ad5ad9689c8f74d8fe71f`.
2. Run `make -j$(nproc)`.

### Expected behavior

_No response_

### Actual behavior

Too long to paste the entire log, so please see https://gist.github.com/mbana/c8e9cad7ef89b580afe48896eb88e34f.

Summary:

```sh
```sh
$ make -j$(nproc)
...
touch -- 'pkg/apis/k0s.k0sproject.io/v1beta1/.client-gen.stamp'
touch -- 'pkg/apis/autopilot.k0sproject.io/v1beta2/.client-gen.stamp'
github.com/k0sproject/k0s/internal/autopilot/testutil imports
	github.com/k0sproject/k0s/pkg/apis/autopilot.k0sproject.io/v1beta2/clientset: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
github.com/k0sproject/k0s/internal/autopilot/testutil imports
	github.com/k0sproject/k0s/pkg/apis/autopilot.k0sproject.io/v1beta2/clientset/fake: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
github.com/k0sproject/k0s/internal/testutil imports
	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/fake: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
github.com/k0sproject/k0s/internal/testutil imports
	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/typed/k0s.k0sproject.io/v1beta1: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
github.com/k0sproject/k0s/pkg/autopilot/controller imports
	github.com/k0sproject/k0s/pkg/apis/autopilot.k0sproject.io/v1beta2/clientset/scheme: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
github.com/k0sproject/k0s/pkg/config imports
	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset: git init --bare in /run/k0s-build/go/mod/cache/vcs/ae919b7ca7cc6029449e16c2a89eb1a5fedae32c585157fb07019da5434cb570: exec: "git": executable file not found in $PATH
make: *** [Makefile:107: go.sum] Error 1
```
```

--- a/.github/workflows/build-airgap-image-bundle.yml
+++ b/.github/workflows/build-airgap-image-bundle.yml
@@ -15,6 +15,9 @@ on:
         description: The airgap image bundle's cache key.
         value: ${{ jobs.build.outputs.cache-key }}
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   build:
     name: "${{ inputs.target-os }}-${{ inputs.target-arch }}"

--- a/.github/workflows/build-docs.yml
+++ b/.github/workflows/build-docs.yml
@@ -11,6 +11,9 @@ on:
       - CODE_OF_CONDUCT.md
       - docs/**
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   build:
     name: Build docs

--- a/.github/workflows/build-k0s.yml
+++ b/.github/workflows/build-k0s.yml
@@ -12,6 +12,9 @@ on:
         required: true
         description: The architecture to build k0s for.
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   build:
     name: ${{ inputs.target-os }}-${{ inputs.target-arch }}

--- a/.github/workflows/check-network.yaml
+++ b/.github/workflows/check-network.yaml
@@ -8,6 +8,9 @@ on:
   schedule:
     - cron: "0 23 * * *"
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   check-network:
     env:

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -28,6 +28,9 @@ on:
       - '.github/workflows/mkdocs-set-default-version.yml'
       - 'mkdocs.yml'
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   prepare:
     name: Prepare

--- a/.github/workflows/lint.yaml
+++ b/.github/workflows/lint.yaml
@@ -29,6 +29,9 @@ on:
       - 'mkdocs.yml'
       - '*.md'
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   lint:
     name: Lint

--- a/.github/workflows/ostests-e2e.yaml
+++ b/.github/workflows/ostests-e2e.yaml
@@ -50,6 +50,7 @@ env:
   # This is hopefully somewhat physically close to where the hosted GitHub runners live.
   # https://github.com/orgs/community/discussions/24969#discussioncomment-3246032
   AWS_REGION: us-east-1
+  MAKEFLAGS: -j
 
 jobs:
   e2e-tests:

--- a/.github/workflows/publish-docs-manual.yml
+++ b/.github/workflows/publish-docs-manual.yml
@@ -7,6 +7,7 @@ on:
         required: true
 
 env:
+  MAKEFLAGS: -j
   PYTHON_VERSION: 3.x
   TARGET_VERSION: ${{ github.event.inputs.version }}
 

--- a/.github/workflows/publish-docs.yml
+++ b/.github/workflows/publish-docs.yml
@@ -7,6 +7,9 @@ on:
     types:
       - published
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   build:
     name: Deploy docs

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -5,6 +5,9 @@ on:
     tags:
       - v* # Push events to matching v*, i.e. v1.0, v20.15.10
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   release:
     env:

--- a/.github/workflows/sbom-upload.yml
+++ b/.github/workflows/sbom-upload.yml
@@ -6,6 +6,9 @@ on:
     branches:
       - main
 
+env:
+  MAKEFLAGS: -j
+
 jobs:
   sbom-upload:
     runs-on: ubuntu-latest

--- a/Makefile
+++ b/Makefile
@@ -126,27 +126,32 @@ pkg/apis/autopilot/v1beta2/.controller-gen.stamp: gen_output_dir = autopilot
 
 pkg/apis/%/.controller-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt hack/tools/Makefile.variables
 	rm -rf 'static/manifests/$(gen_output_dir)/CustomResourceDefinition'
-	rm -f -- '$(dir $@)'zz_*.go
-	CGO_ENABLED=0 $(GO) run sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version) \
-	  crd \
-	  paths="./$(dir $@)..." \
-	  output:crd:artifacts:config=./static/manifests/$(gen_output_dir)/CustomResourceDefinition \
-	  object:headerFile=hack/tools/boilerplate.go.txt
+	mkdir -p 'static/manifests/$(gen_output_dir)'
+	gendir="$$(mktemp -d .controller-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version) \
+	    paths="./$(dir $@)..." \
+	    object:headerFile=hack/tools/boilerplate.go.txt output:object:dir="$$gendir" \
+	    crd output:crd:dir='static/manifests/$(gen_output_dir)/CustomResourceDefinition' \
+	  && mv -f -- "$$gendir"/zz_generated.deepcopy.go '$(dir $@).'
 	touch -- '$@'
 
 clientset_input_dirs := pkg/apis/autopilot/v1beta2 pkg/apis/k0s/v1beta1 pkg/apis/helm/v1beta1
 codegen_targets += pkg/client/clientset/.client-gen.stamp
 pkg/client/clientset/.client-gen.stamp: $(shell find $(clientset_input_dirs) -type f -name \*.go -not -name \*_test.go -not -name zz_\*)
 pkg/client/clientset/.client-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt embedded-bins/Makefile.variables
-	rm -rf -- '$(dir $@)'
-	CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
-	  --go-header-file=hack/tools/boilerplate.go.txt \
-	  --input-base='' \
-	  --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
-	  --output-package=github.com/k0sproject/k0s/$(patsubst %/,%,$(dir $(patsubst %/,%,$(dir $@)))) \
-	  --clientset-name=$(notdir $(patsubst %/,%,$(dir $@))) \
-	  --output-base=. \
-	  --trim-path-prefix=github.com/k0sproject/k0s
+	gendir="$$(mktemp -d .client-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
+	    --go-header-file=hack/tools/boilerplate.go.txt \
+	    --input-base='' \
+	    --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
+	    --output-package=github.com/k0sproject/k0s/pkg/client \
+	    --clientset-name=clientset \
+	    --output-base="$$gendir/out" \
+	    --trim-path-prefix=github.com/k0sproject/k0s \
+	  && { [ ! -e pkg/client/clientset ] || mv -- pkg/client/clientset "$$gendir/old"; } \
+	  && mv -f -- "$$gendir/out/github.com/k0sproject/k0s/pkg/client/clientset" pkg/client/.
 	touch -- '$@'
 
 codegen_targets += static/zz_generated_assets.go

--- a/Makefile
+++ b/Makefile
@@ -112,18 +112,19 @@ $(K0S_GO_BUILD_CACHE):
 go.sum: go.mod .k0sbuild.docker-image.k0s
 	$(GO) mod tidy && touch -c -- '$@'
 
-codegen_targets += pkg/apis/helm/v1beta1/.controller-gen.stamp
+controllergen_targets += pkg/apis/helm/v1beta1/.controller-gen.stamp
 pkg/apis/helm/v1beta1/.controller-gen.stamp: $(shell find pkg/apis/helm/v1beta1/ -maxdepth 1 -type f -name \*.go)
 pkg/apis/helm/v1beta1/.controller-gen.stamp: gen_output_dir = helm
 
-codegen_targets += pkg/apis/k0s/v1beta1/.controller-gen.stamp
+controllergen_targets += pkg/apis/k0s/v1beta1/.controller-gen.stamp
 pkg/apis/k0s/v1beta1/.controller-gen.stamp: $(shell find pkg/apis/k0s/v1beta1/ -maxdepth 1 -type f -name \*.go)
 pkg/apis/k0s/v1beta1/.controller-gen.stamp: gen_output_dir = v1beta1
 
-codegen_targets += pkg/apis/autopilot/v1beta2/.controller-gen.stamp
+controllergen_targets += pkg/apis/autopilot/v1beta2/.controller-gen.stamp
 pkg/apis/autopilot/v1beta2/.controller-gen.stamp: $(shell find pkg/apis/autopilot/v1beta2/ -maxdepth 1 -type f -name \*.go)
 pkg/apis/autopilot/v1beta2/.controller-gen.stamp: gen_output_dir = autopilot
 
+codegen_targets += $(controllergen_targets)
 pkg/apis/%/.controller-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt hack/tools/Makefile.variables
 	rm -rf 'static/manifests/$(gen_output_dir)/CustomResourceDefinition'
 	mkdir -p 'static/manifests/$(gen_output_dir)'
@@ -155,12 +156,17 @@ pkg/client/clientset/.client-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/bo
 	touch -- '$@'
 
 codegen_targets += static/zz_generated_assets.go
-static_asset_dirs := static/manifests static/misc
+static/zz_generated_assets.go: $(controllergen_targets) # to generate the CRDs into static/manifests/*/CustomResourceDefinition
+static/zz_generated_assets.go: $(shell find static/manifests/calico static/manifests/windows static/misc -type f)
 static/zz_generated_assets.go: .k0sbuild.docker-image.k0s hack/tools/Makefile.variables
-static/zz_generated_assets.go: $(shell find $(static_asset_dirs) -type f)
-	-rm -f -- '$@'
-	CGO_ENABLED=0 $(GO) install github.com/kevinburke/go-bindata/go-bindata@v$(go-bindata_version)
-	$(GO_ENV) go-bindata -o '$@' -pkg static -prefix static $(patsubst %,%/...,$(static_asset_dirs))
+	CGO_ENABLED=0 $(GO) run github.com/kevinburke/go-bindata/go-bindata@v$(go-bindata_version) \
+	  -o '$@' -pkg static -prefix static \
+	  static/manifests/helm/CustomResourceDefinition/... \
+	  static/manifests/v1beta1/CustomResourceDefinition/... \
+	  static/manifests/autopilot/CustomResourceDefinition/... \
+	  static/manifests/calico/... \
+	  static/manifests/windows/... \
+	  static/misc/...
 
 ifeq ($(EMBEDDED_BINS_BUILDMODE),none)
 BUILD_GO_TAGS += noembedbins

--- a/embedded-bins/Makefile
+++ b/embedded-bins/Makefile
@@ -13,7 +13,6 @@ SOURCE_DATE_EPOCH ?= $(shell git log -1 --pretty=%ct || date -u +%s)
 bindir = staging/${TARGET_OS}/bin
 posix_bins = runc kubelet containerd containerd-shim containerd-shim-runc-v1 containerd-shim-runc-v2 kube-apiserver kube-scheduler kube-controller-manager etcd kine konnectivity-server xtables-legacy-multi xtables-nft-multi
 windows_bins = kubelet.exe kube-proxy.exe containerd.exe containerd-shim-runhcs-v1.exe
-buildmode = docker
 
 ifeq ($(TARGET_OS),windows)
 bins = ${windows_bins}

--- a/embedded-bins/Makefile
+++ b/embedded-bins/Makefile
@@ -65,7 +65,12 @@ $(bindir)/kubelet.exe $(bindir)/kube-proxy.exe: .container.kubernetes.windows
 $(bindir)/containerd.exe $(bindir)/containerd-shim-runhcs-v1.exe: .container.containerd.windows
 
 $(addprefix $(bindir)/, $(bins)): | $(bindir)
-	docker export $$(cat $<) | tar -C $(dir $(bindir)) -xv bin/$(notdir $@) && touch $@
+	tardir=$$(mktemp -d -- '$(notdir $@).XXXXXX.tmp') \
+	  && trap "rm -rf -- $$tardir" INT EXIT \
+	  && docker export "$$(cat $<)" \
+	  | tar -C "$$tardir" -x bin/$(notdir $@) \
+	  && touch -c -- "$$tardir"/bin/$(notdir $@) \
+	  && mv "$$tardir"/bin/$(notdir $@) $@
 
 build_docker_container = \
 	rm -f -- '$@'; \

--- a/vars.sh
+++ b/vars.sh
@@ -3,6 +3,9 @@
 
 set -eu
 
+# Don't deal with recursive make stuff here!
+unset -v MAKEFLAGS MAKELEVEL
+
 from=
 var=
 

